### DISCUSSION TITLE
wlr/taskbar: List windows in correct order?

### DISCUSSION THREAD

**Discussion by peret** [YaLTeR/niri#549] [Ideas]
Hello,

first of all, I _think_ this is the proper repo to raise this issue in, but please let me know in case I'm wrong and these changes would need to be addressed in waybar or wlr/taskbar.

So, I configured my waybar with the wlr/taskbar module to display a list of icons of my open windows. It works, but there are two changes I would wish for, that would make it way more useful for me to actually navigate windows.

1. I wish the order of applications in the taskbar would reflect the actual order of windows in the current workspace. Right now, I think they are ordered by when they were started. So, if I reorder the windows in my workspace, the taskbar does not reflect that change. It can be configured to always show the active window first, but that doesn't change the rest of the order.
I think this change would make it easier to know at a glance how to navigate to a window/application I'm looking for.
2. I wish the taskbar would only show application windows of the current workspace, for the same reason. Also, I guess this change would be necessary to implement 1., since otherwise it would be unclear how to sort applications on different workspaces, relative to each other.

My understanding is that this behavior is controlled by niri's implementation of the wlr-foreign-toplevel-management API, but I don't know enough about this API yet. So first, would it be possible to make these changes in niri? Second, do y'all think this would be good changes to make? I realize that this might not be the behavior that every user expects.
If so, I might attempt a PR, but I wanted to make sure that it'd be possible, first :D

**Comment by YaLTeR** [comment]
> I wish the order of applications in the taskbar would reflect the actual order of windows in the current workspace.

As far as I understand the wlr-foreign-toplevel-management protocol, there's no defined window ordering. So Waybar happens to show windows in *some* order and niri happens to send windows in *some* order and there's no defined way for niri to send windows in a specific order, or rearrange them after the fact.

> I wish the taskbar would only show application windows of the current workspace

There's no Wayland protocol that communicates the concept of workspaces. There's https://gitlab.freedesktop.org/wayland/wayland-protocols/-/merge_requests/40 that adds that but it is in a bit of a limbo.

There's also #453 which will allow doing something like this in a niri-specific fashion.

**Reply by 3ulalia** [reply]
Now that the event stream IPC has been merged, I've been looking at implementing a `niri/taskbar` module in waybar that would behave in the ways described above. However, I don't think Niri currently communicates window position in the IPC (or anywhere, really). Should I make an issue/PR to track this?

edit: ah i see https://github.com/YaLTeR/niri/discussions/624

**Reply by YaLTeR** [reply]
Yeah, that issue also has some of my thoughts on how this should be exposed.
